### PR TITLE
SITL running on Webots

### PR DESCRIPTION
**Pull request by HefnySco**
Webots is an open source robot simulator that provides a complete development environment to model, program and simulate robots. 

The main important part in the pull request is the controller file that reads data from webots world and send it to ardupilot sitl. the ardupilot sitl module is a replication of SIM_Morse.cpp but slightly modified for Webots.

The module is working but needs enhancements in PID params and drone structure design.

### CODE REVIEW

**Comment by HefnySco** [comment]
This is a demo [https://youtu.be/Z_BEDswHFls](https://youtu.be/Z_BEDswHFls)  as you can see there are issues in YAW. and the simulation is very fragile to changes in values in PID. on the otherside sensors scales and axis direction are all correct -I hope :) as I did alot of tests-.

**Comment by tridge** [comment]
really nice! The commits need a bit of cleanup to meet our coding conventions (I can do that for you if you like)
Do you have a log of it flying? I'd be interested to check if the EKF innovations are low, to demonstrate that the lock-step scheduling is working correctly

**Comment by HefnySco** [comment]
Thank you @tridge 
**> (I can do that for you if you like)** 
Please do or guide me through it. Thank you.
**> Do you have a log of it flying?**
Kindly find attached log. [log_0_٢٠١٩-٧-٣١-١٣-١٢-٥٢.zip](https://github.com/ArduPilot/ardupilot/files/3451439/log_0_.-.-.-.-.-.zip)
yes, I believe that some of the issues is due to timing between webots and SITL as sometimes it works well and othertimes it fails without changing any numbers, especially when I use OBS software to video capture my screen.

**Comment by HefnySco** [comment]
This is https://docs.google.com/spreadsheets/d/1ZXIDLAhBHd2HfScILtGQehjGkqyf3NcTf5lo6G59MAk/edit?usp=sharing the calculation I used to calculate Thrust and make it linear with input.
I am not sure how to properly scale Torque values so I used try & error with torque which might affect YAW_PID and other parameters.
My objective to keep PIDs realistic not just flyable.

**Comment by tridge** [comment]
> Please do or guide me through it. Thank you.

I've done it and pushed to your branch. Note that it is a good idea to use a branch name other than 'master'. For example 'pr-webots' would have been good. 

> Kindly find attached log. [log_0_٢٠١٩-٧-٣١-١٣-١٢-٥٢.zip](https://github.com/ArduPilot/ardupilot/files/3451439/log_0_.-.-.-.-.-.zip)

the EKF innovations in that are quite low, so I think you are doing well with the kinematic consistency

**Comment by HefnySco** [comment]
I added a branch [https://github.com/HefnySco/ardupilot/tree/pr-webots/libraries](https://github.com/HefnySco/ardupilot/tree/pr-webots/libraries)

but I hope I didnt make stupid thing because I could not find changes you made. -sorry I am bad in git-.

**Comment by HefnySco** [comment]
> > Please do or guide me through it. Thank you.
> 
> I've done it and pushed to your branch. Note that it is a good idea to use a branch name other than 'master'. For example 'pr-webots' would have been good.

do you mean this file libraries/AP_HAL_SITL/SITL_cmdline.cpp ? 
I missed it during the first upload :( 

Thank you very much @tridge

**Comment by HefnySco** [comment]
> I've done it and pushed to your branch. Note that it is a good idea to use a branch name other than 'master'. For example 'pr-webots' would have been good.
> 
> > Kindly find attached log. [log_0_٢٠١٩-٧-٣١-١٣-١٢-٥٢.zip](https://github.com/ArduPilot/ardupilot/files/3451439/log_0_.-.-.-.-.-.zip)
> 
> **the EKF innovations in that are quite low, so I think you are doing well with the kinematic consistency**
Now regarding this point. I tried to use **SIM_SPEEDUP  0.1** and slow down "basicTimeStep" in webots but that made very bad results.

Is there a way to slow down simulators so that I can guarantee that it works fine on slower computers ?

kindly advise.

**Comment by HefnySco** [comment]
> Is there a way to slow down simulators so that I can guarantee that it works fine on slower computers ?

I followed this [link](https://www.cyberbotics.com/doc/guide/speed-performance) and performance is much better.

File: libraries/SITL/SIM_Webots.cpp [hunk 1]
```diff
@@ -0,0 +1,581 @@
...
+    sim_sock->send(buf, strlen(buf));
+}
+
+
+/*
+  update the Webots simulation by one time step
+ */
+void Webots::update(const struct sitl_input &input)
+{   
+    #define M_PI2 6.28318530718
```
Comment (khancyr) [inline libraries/SITL/SIM_Webots.cpp]: not used

Comment (HefnySco) [reply libraries/SITL/SIM_Webots.cpp]: I have removed them.
Thank you :)

**Comment by HefnySco** [comment]
Now the simulator has Quad in Plus and X modes and Rover.

**Comment by HefnySco** [comment]
@tridge @OXINARF I closed this pr and opened a new one from a branch pr-webots on my cloned repository.

**Comment by HefnySco** [comment]
The same PR is here [https://github.com/ArduPilot/ardupilot/pull/12006](https://github.com/ArduPilot/ardupilot/pull/12006)

**Pull request closed without merging**